doc/chapters/debugging/index.rst

```diff
@@ -0,0 +1,197 @@
+
+Debugging applications running in containers
+============================================
+
+This section and subsections describes some common ways of debugging
+applications running inside containers and how to use common tools on the
+applications running inside the containers.
+
+LXC
+---
+
+LXC provides several tools that can be used to check out the containers
+currently running on the system. `Note that all commands must be run as the main
+user which the containers are started as, ie root usually.`
+
+*lxc-ls* lists all currently running containers on the host.
+
+*lxc-attach* can be used to "get inside" the container with a shell for example,
+or to test run commands on the inside of the container. Please note that the
+command given must be available inside the container.
+
+*lxc-console* can be used to login with a shell inside the container in question.
+
+*lxc-info* provides information about the running container, such as network
+interface used for t he bridge, the state of the container, what PID it has and
+so forth.
+
+Cleanup
+-------
+
+The internal filesystem will always be deleted and recreated when an app is
+restarted. This means that anything you want to save during debugging needs to
+be saved to a bind mounted directory before you shut down the container.
+
+`SoftwareContainer` will also delete all container instances upon startup,
+meaning that container instances will be lost when restarting.
+
+GDB
+---
+
+Attaching a `gdb` session can be done in two ways, one is to run the `gdb` session
+and attach inside the container, the other is to attach to the process from the
+outside of the container. Using both methods have their own set of problems you
+need to solve.
+
+Running GDB inside container
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you run `gdb` from inside the container using `lxc-attach`, you should not
+need to change any paths etc to find the correct libraries etc. You will need
+to have the `PID` from inside the container available to connect to however. this
+can be found out by running::
+
+    sudo lxc-attach --name SC-X ps aux
+
+Then running::
+
+    sudo lxc-attach --name SC-X -- gdb --pid YY
+
+Replace YY with the `PID` as seen inside the container. You can also run the
+application directly inside the container and start it with gdb directly::
+
+    sudo lxc-attach --name SC-X -- gdb --args /bin/ls
+
+TODO: Uncertain why, but this fails with unable to find executable.
+
+Running GDB outside container
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you run `gdb` from outside the container you will run into some problems with the
+paths as the application is executing inside a `chroot` and `gdb` is executing
+outside of it, this is fairly easily fixed however. First off you need to
+provide `gdb` with the path of the executable as the path provided by the system
+is inside the chroot::
+
+    # ps aux |grep main
+    # sudo gdb --pid XXX /tmp/container//SC-0/gateways/app/main
```

Comment:
I mean that one can bind-mount to any location in the container, in which case we first mount into /gateways, and then move the mount (to some place which the host can't easily access). Will the gdb chroot trick still work then?